### ISSUE TITLE
Explicit Built-in versioning

### ISSUE THREAD

**Issue by matheus23** [dhall-lang/dhall-lang#835]
One thing I like about Dhall is that it is pretty clear what your configuration depends on, as there are no hidden mutable state dependencies and dependencies on the environment or other dhall files can be statically traced.

However, when looking at a dhall file, it is not clear which version of dhall built-ins is expected.
I want to discuss one way of being explicit about built-ins: Parameterizing dhall files with a record containing all built-ins which would be automatically provided by dhall implementations:

```dhall
  λ ( dhall : https://v11.dhall-lang.org/ sha256:0c3c40a63108f2e6ad59f23b789c18eb484d0e9aebc9416c5a4f338c6753084b )
→ -- use dhall.Text, dhall.Bool, dhall.Natural/isZero, etc.
```

And `https://v11.dhall-lang.org/` would resolve to:
```dhall
{ Bool : Type
, Natural : Type
, Text : Type
, List : Type -> Type
, Optional : Type -> Type
, Integer : Type
, Natural/build : (∀(natural : Type) → ∀(succ : natural → natural) → ∀(zero : natural) → natural) → Natural
, Natural/fold : Natural → ∀(natural : Type) → ∀(succ : natural → natural) → ∀(zero : natural) → natural
, Natural/isZero : Natural → Bool
, Natural/even : Natural → Bool
, Natural/odd : Natural → Bool
, Natural/toInteger : Natural → Integer
, Natural/show : Natural → Text
, Natural/subtract : Natural → Natural → Natural
, Text/show : Text → Text
, List/build : ∀(a : Type) → (∀(list : Type) → ∀(cons : a → list → list) → ∀(nil : list) → list) → List a
, List/fold : ∀(a : Type) → List a → ∀(list : Type) → ∀(cons : a → list → list) → ∀(nil : list) → list
, List/length : ∀(a : Type) → List a → Natural
, List/head : ∀(a : Type) → List a → Optional a
, List/last : ∀(a : Type) → List a → Optional a
, List/indexed : ∀(a : Type) → List a → List { index : Natural, value : a }
, List/reverse : ∀(a : Type) → List a → List a
, Optional/fold : ∀(a : Type) → Optional a → ∀(optional : Type) → ∀(just : a → optional) → ∀(nothing : optional) → optional
, Optional/build : ∀(a : Type) → (∀(optional : Type) → ∀(just : a → optional) → ∀(nothing : optional) → optional) → Optional a
, Integer/show : Integer → Text
, Integer/toDouble : Integer → Double
, Integer/negate : Integer → Integer
, Integer/clamp : Integer → Natural
, Double/show : Double → Text
}
```

It has some benefits:
* The dhall builtin versions are now explicit
* Builtins in the standard that can be represented as a type inside dhall (so, unfortunately not `/\` or `assert`) can be added/removed, while old dhall code can still be run using functions that convert between dhall versions e.g. V11->V10 for compatibility, for example, imagine `List.indexed` would be removed in `V13`, but new primitives would allow a function of type `V13 -> V12`. 
* Dhall implementations could experiment with alternative sets of builtins without breaking compatibility (just match the type of the dhall file to the your available builtins).

But it is obviously a trade-off, so some drawbacks are:
* Increased complexity:
  - Every dhall file would become a function.
  - Every import would need to be applied to the current builtins (or a compatiblised version)
  - Every builtin would become a qualified identifier: `Text -> dhall.Text`
* No support for including operators like `==` or `#` in the builtin type, as we don't have support for infix operators, yet

Above points show, that this is not anything that can be implemented right now. I'm just curious what your thoughts on this are, and whether steering towards this as a goal would make sense? I feel like many drawbacks could be solved with changes to the syntax.

**Comment by Nadrieril** [comment]
Dhall also regularly has breaking changes in the syntax and the normalization semantics. I feel that a general way of indicating the supported dhall versions would be amazingly helpful; that way a given implementation could be free to support a range of versions of the standard without breaking anything for older files

**Comment by ari-becker** [comment]
From what I understand, the current idiomatic approach is to use a top-level `imports.dhall` file, e.g. https://github.com/coralogix/dhall-kops/blob/master/imports.dhall which is imported by all other Dhall expressions in the same repository. This way, the supported version is explicit, and migrating between Dhall versions for a non-trivial number of Dhall expressions doesn't take a million years.

**Comment by Gabriel439** [comment]
Yeah, I like @ari-becker's idea of having a package that serves as a version compatibility shim

**Comment by matheus23** [comment]
I noticed #800 also mentions exactly what I was thinking, but addresses these points:

> Unknown-sized annoyance at random times is worse than one predictable constant-time annoyance, so you might create BuiltinsType.dhall and institute a rule that all modules shall begin with λ(builtins : ./BuiltinsType.dhall) →.
> 
> This is the same thing but with worse ergonomics. Plus it's something that has to be discovered (as an idiom) and maintained (as a styleguide rule).

However, this proposal wouldn't suffer the discoverability/styleguide issue.

I actually think this is a clean solution (for builtin versioning and specifying what builtins should be available!), it doesn't even need any language syntax/semantics changes and I think it will make the standard simpler in the future.

**Comment by Gabriel439** [comment]
@matheus23: I think requiring every Dhall file to be a function is too large of a negative, for two reasons:

* It is disruptive to existing Dhall code
* It deteriorates the ergonomics of Dhall's import system

I think #800 could still address your use case, because with forward declarations we could do something similar to what you suggest, like this:

```dhall
let builtins : https://v11.dhall-lang.org/ sha256:0c3c40a63108f2e6ad59f23b789c18eb484d0e9aebc9416c5a4f338c6753084b

in  …
```

**Comment by sjakobi** [comment]
In Haskell I often have the same problem that I don't know which GHC versions some code is going to compile with. I usually figure it out by just trying different GHC versions.

While it's not an elegant solution by any means, I wonder whether `dhall` should make it easier to use multiple versions by shipping symbolic links like `dhall-1.27.0`.

A `dhall standard-version` command might also make it easier to make the connection from a version of the `dhall` executable to the supported version of the Dhall standard.
